File: packages/typescript/service/src/hooks.ts
/**
 * ANIP v0.11 — Observability hook type definitions.
 *
 * Callback-based injection points for logging, metrics, tracing, and
 * diagnostics.  Every callback is optional; absent hooks are silent.
 */

// ---------------------------------------------------------------------------
// hooks.logging — Structured Logging Hooks
// ---------------------------------------------------------------------------

export interface LoggingHooks {
  onInvocationStart?(event: {
    capability: string;
    invocationId: string;
    clientReferenceId: string | null;
    rootPrincipal: string;
    subject: string;
    timestamp: string;
  }): void;

  onInvocationEnd?(event: {
    capability: string;
    invocationId: string;
    success: boolean;
    failureType: string | null;
    durationMs: number;
    timestamp: string;
  }): void;

  onDelegationFailure?(event: {
    reason: string;
    tokenId: string | null;
    timestamp: string;
  }): void;

  onAuditAppend?(event: {
    sequenceNumber: number;
    capability: string;
    invocationId: string;
    success: boolean;
    timestamp: string;
  }): void;

  onCheckpointCreated?(event: {
    checkpointId: string;
    entryCount: number;
    merkleRoot: string;
    timestamp: string;
  }): void;

  onRetentionSweep?(event: {
    deletedCount: number;
    durationMs: number;
    timestamp: string;
  }): void;

  onAggregationFlush?(event: {
    windowCount: number;
    entriesFlushed: number;
    timestamp: string;
  }): void;

  onStreamingSummary?(event: {
    invocationId: string;
    capability: string;
    eventsEmitted: number;
    eventsDelivered: number;
    clientDisconnected: boolean;
    durationMs: number;
    timestamp: string;
  }): void;
}

// ---------------------------------------------------------------------------
// hooks.metrics — Metrics Hooks
// ---------------------------------------------------------------------------

export interface MetricsHooks {
  onInvocationDuration?(event: {
    capability: string;
    durationMs: number;
    success: boolean;
  }): void;

  onDelegationDenied?(event: { reason: string }): void;

  onAuditAppendDuration?(event: {
    durationMs: number;
    success: boolean;
  }): void;

  onCheckpointCreated?(event: { lagSeconds: number }): void;

  onCheckpointFailed?(event: { error: string }): void;

  onProofGenerated?(event: { durationMs: number }): void;

  onProofUnavailable?(event: { reason: string }): void;

  onRetentionDeleted?(event: { count: number }): void;

  onAggregationFlushed?(event: { windowCount: number }): void;

  onStreamingDeliveryFailure?(event: { capability: string }): void;
}

// ---------------------------------------------------------------------------
// hooks.tracing — Tracing Hooks
// ---------------------------------------------------------------------------

export interface TracingHooks {
  startSpan?(event: {
    name: string;
    attributes: Record<string, string | number | boolean>;
    parentSpan?: unknown;
  }): unknown;

  endSpan?(event: {
    span: unknown;
    status: "ok" | "error";
    errorType?: string;
    errorMessage?: string;
    attributes?: Record<string, string | number | boolean>;
  }): void;
}

// ---------------------------------------------------------------------------
// hooks.diagnostics — Background Failure Signals
// ---------------------------------------------------------------------------

export interface DiagnosticsHooks {
  onBackgroundError?(event: {
    source: "checkpoint" | "retention" | "aggregation";
    error: string;
    timestamp: string;
  }): void;
}

// ---------------------------------------------------------------------------
// ANIPHooks — Top-level container
// ---------------------------------------------------------------------------

export interface ANIPHooks {
  logging?: LoggingHooks;
  metrics?: MetricsHooks;
  tracing?: TracingHooks;
  diagnostics?: DiagnosticsHooks;
}

// ---------------------------------------------------------------------------
// HealthReport — Pull-based diagnostics
// ---------------------------------------------------------------------------

export interface HealthReport {
  status: "healthy" | "degraded" | "unhealthy";
  storage: { type: string };
  checkpoint: {
    healthy: boolean;
    lastRunAt: string | null;
    lagSeconds: number | null;
  } | null;
  retention: {
    healthy: boolean;
    lastRunAt: string | null;
    lastDeletedCount: number;
  };
  aggregation: { pendingWindows: number } | null;
}

File: packages/python/anip-service/src/anip_service/hooks.py
"""Observability hook types for the ANIP service runtime."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable


@dataclass
class LoggingHooks:
    on_invocation_start: Callable[[dict[str, Any]], None] | None = None
    on_invocation_end: Callable[[dict[str, Any]], None] | None = None
    on_delegation_failure: Callable[[dict[str, Any]], None] | None = None
    on_audit_append: Callable[[dict[str, Any]], None] | None = None
    on_checkpoint_created: Callable[[dict[str, Any]], None] | None = None
    on_retention_sweep: Callable[[dict[str, Any]], None] | None = None
    on_aggregation_flush: Callable[[dict[str, Any]], None] | None = None
    on_streaming_summary: Callable[[dict[str, Any]], None] | None = None


@dataclass
class MetricsHooks:
    on_invocation_duration: Callable[[dict[str, Any]], None] | None = None
    on_delegation_denied: Callable[[dict[str, Any]], None] | None = None
    on_audit_append_duration: Callable[[dict[str, Any]], None] | None = None
    on_checkpoint_created: Callable[[dict[str, Any]], None] | None = None
    on_checkpoint_failed: Callable[[dict[str, Any]], None] | None = None
    on_proof_generated: Callable[[dict[str, Any]], None] | None = None
    on_proof_unavailable: Callable[[dict[str, Any]], None] | None = None
    on_retention_deleted: Callable[[dict[str, Any]], None] | None = None
    on_aggregation_flushed: Callable[[dict[str, Any]], None] | None = None
    on_streaming_delivery_failure: Callable[[dict[str, Any]], None] | None = None


@dataclass
class TracingHooks:
    start_span: Callable[[dict[str, Any]], Any] | None = None
    end_span: Callable[[dict[str, Any]], None] | None = None


@dataclass
class DiagnosticsHooks:
    on_background_error: Callable[[dict[str, Any]], None] | None = None


@dataclass
class ANIPHooks:
    logging: LoggingHooks | None = None
    metrics: MetricsHooks | None = None
    tracing: TracingHooks | None = None
    diagnostics: DiagnosticsHooks | None = None


@dataclass
class HealthReport:
    status: str  # "healthy" | "degraded" | "unhealthy"
    storage: dict[str, Any]
    checkpoint: dict[str, Any] | None
    retention: dict[str, Any]
    aggregation: dict[str, Any] | None
